File: scripts/get_last_cint_payloads.js
#!/usr/bin/env node
/**
 * Get Last N Cint Respondent Payloads (JavaScript version)
 * 
 * This script connects to MongoDB and retrieves the last N respondent payloads
 * that were sent to Cint.
 * 
 * Usage:
 *   node scripts/get_last_cint_payloads.js [limit]
 *   
 * Examples:
 *   node scripts/get_last_cint_payloads.js           # Get last 10
 *   node scripts/get_last_cint_payloads.js 20        # Get last 20
 */

const { MongoClient, ObjectId } = require('mongodb');

const MONGO_URI = process.env.MONGO_URI || 'mongodb://localhost:27017/';
const CINT_SUPPLIER_CODE = process.env.CINT_SUPPLIER_CODE || '6777';
const CINT_CALLBACK_URL = process.env.CINT_CALLBACK_URL || 'https://torpedo.cogentixresearch.com/api/cint/status';

function formatTimestamp(date) {
    if (!date) return 'N/A';
    return date.toISOString().replace('T', ' ').split('.')[0] + ' UTC';
}

async function getLastCintPayloads(limit = 10) {
    const client = new MongoClient(MONGO_URI);
    
    try {
        await client.connect();
        console.log('✓ Connected to MongoDB');
        
        // Access databases
        const surveyDb = client.db('survey_allocation');
        const allocationLog = surveyDb.collection('allocation_log');
        const surveysCollection = surveyDb.collection('surveys');
        const respondentsCollection = surveyDb.collection('respondents');
        
        const cintDb = client.db('cint_research');
        const cintSurveys = cintDb.collection('cint_surveys');
        
        const payloads = [];
        
        // Use aggregation pipeline to efficiently filter for CINT allocations
        // This joins with surveys collection and filters in the database
        const pipeline = [
            // Sort by timestamp descending (most recent first)
            { $sort: { timestamp: -1 } },
            
            // Join with surveys collection to get provider info
            { $lookup: {
                from: 'surveys',
                localField: 'survey_id',
                foreignField: '_id',
                as: 'survey'
            }},
            
            // Unwind the survey array
            { $unwind: { path: '$survey', preserveNullAndEmptyArrays: false } },
            
            // Filter for CINT provider only
            { $match: { 'survey.provider': 'CINT' } },
            
            // Limit to requested number of results
            { $limit: limit }
        ];
        
        const cintLogs = await allocationLog.aggregate(pipeline).toArray();
        
        // Build detailed payload information for each log
        for (const log of cintLogs) {
            const surveyId = log.survey_id;
            const respondentId = log.rid || log.respondent_id;
            
            // Get survey details from the joined data
            const survey = log.survey;
            const externalSurveyId = survey?.external_id;
            const surveyName = survey?.name;
            
            // Get respondent details
            const respondent = await respondentsCollection.findOne({ rid: respondentId });
            
            // Try to get Cint survey details
            let cintSurvey = null;
            if (externalSurveyId) {
                cintSurvey = await cintSurveys.findOne({ survey_id: parseInt(externalSurveyId) });
            }
            
            // Reconstruct the payload that was sent to Cint
            const payload = {
                survey_id: externalSurveyId ? String(externalSurveyId) : String(surveyId),
                supplier_code: CINT_SUPPLIER_CODE,
                respondent_id: respondentId,
                secure_hash: '[HMAC-SHA256 hash - not stored]',
                return_url: CINT_CALLBACK_URL
            };
            
            // Additional metadata
            const metadata = {
                timestamp: formatTimestamp(log.timestamp),
                allocation_id: log.allocation_id ? String(log.allocation_id) : '',
                vid: log.vid,
                cc: log.cc,
                ip_address: log.ip_address,
                user_agent: log.user_agent,
                survey_name: surveyName,
                survey_status: survey?.status
            };
            
            const fullRecord = {
                payload_sent_to_cint: payload,
                metadata: metadata
            };
            
            // Add Cint survey details if available
            if (cintSurvey) {
                fullRecord.cint_survey_details = {
                    survey_name: cintSurvey.survey_name,
                    country_language: cintSurvey.country_language,
                    loi: cintSurvey.bid_length_of_interview,
                    cpi: cintSurvey.revenue_per_interview?.value,
                    conversion: cintSurvey.conversion
                };
            }
            
            payloads.push(fullRecord);
        }
        
        return payloads;
        
    } catch (error) {
        console.error('Error retrieving payloads:', error);
        throw error;
    } finally {
        await client.close();
    }
}

function printPayloads(payloads) {
    if (!payloads || payloads.length === 0) {
        console.log('\nNo Cint payloads found.');
        return;
    }
    
    console.log('\n' + '='.repeat(80));
    console.log(`LAST ${payloads.length} RESPONDENT PAYLOADS SENT TO CINT`);
    console.log('='.repeat(80));
    console.log();
    
    payloads.forEach((record, i) => {
        console.log(`[${i + 1}] Timestamp: ${record.metadata.timestamp}`);
        console.log('-'.repeat(80));
        
        console.log('\nPayload sent to Cint API:');
        console.log('  POST https://api.samplicio.us/supply/v1/entrylinks');
        const payload = record.payload_sent_to_cint;
        for (const [key, value] of Object.entries(payload)) {
            console.log(`    ${key}: ${value}`);
        }
        
        console.log('\nMetadata:');
        const metadata = record.metadata;
        for (const [key, value] of Object.entries(metadata)) {
            if (key !== 'timestamp') {  // Already shown above
                console.log(`    ${key}: ${value || 'N/A'}`);
            }
        }
        
        if (record.cint_survey_details) {
            console.log('\nCint Survey Details:');
            const details = record.cint_survey_details;
            for (const [key, value] of Object.entries(details)) {
                console.log(`    ${key}: ${value || 'N/A'}`);
            }
        }
        
        console.log('\n' + '='.repeat(80));
        console.log();
    });
}

async function main() {
    const limit = parseInt(process.argv[2]) || 10;
    
    console.log('Connecting to MongoDB...');
    console.log(`Retrieving last ${limit} Cint respondent payloads...\n`);
    
    try {
        const payloads = await getLastCintPayloads(limit);
        printPayloads(payloads);
        console.log(`\nTotal payloads retrieved: ${payloads.length}`);
        process.exit(0);
    } catch (error) {
        console.error('\nError:', error.message);
        process.exit(1);
    }
}

// Run if called directly
if (require.main === module) {
    main();
}

module.exports = { getLastCintPayloads };

File: scripts/get_last_cint_payloads.py
#!/usr/bin/env python3
"""
Get Last N Cint Respondent Payloads

This script connects to MongoDB and retrieves the last N respondent payloads
that were sent to Cint. It reconstructs the full payload data including:
- survey_id
- supplier_code  
- respondent_id
- secure_hash (if available)
- return_url
- timestamp
- allocation details

Usage:
    python scripts/get_last_cint_payloads.py
    python scripts/get_last_cint_payloads.py --limit 20
    python scripts/get_last_cint_payloads.py --json output.json
    
Or on VM via SSH:
    ssh user@vm "cd /path/to/campaign_platform && python scripts/get_last_cint_payloads.py"
"""

import os
import sys
from pymongo import MongoClient
from datetime import datetime
from typing import List, Dict, Any
import json

# Add parent directory to path for imports
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))

def get_mongo_uri() -> str:
    """Get MongoDB URI from environment or use default"""
    return os.getenv('MONGO_URI', 'mongodb://localhost:27017/')

def format_timestamp(dt) -> str:
    """Format datetime to readable string"""
    if isinstance(dt, datetime):
        return dt.strftime('%Y-%m-%d %H:%M:%S UTC')
    return str(dt)

def get_last_cint_payloads(limit: int = 10) -> List[Dict[str, Any]]:
    """
    Retrieve last N respondent payloads sent to Cint
    
    Args:
        limit: Number of records to retrieve (default: 10)
        
    Returns:
        List of payload dictionaries with full details
    """
    mongo_uri = get_mongo_uri()
    client = MongoClient(mongo_uri)
    
    # Access survey_allocation database
    survey_db = client['survey_allocation']
    allocation_log = survey_db['allocation_log']
    surveys_collection = survey_db['surveys']
    respondents_collection = survey_db['respondents']
    
    # Access cint_research database for additional Cint-specific data
    cint_db = client['cint_research']
    cint_surveys = cint_db['cint_surveys']
    
    payloads = []
    
    try:
        # Use aggregation pipeline to efficiently filter for CINT allocations
        # This joins with surveys collection and filters in the database
        pipeline = [
            # Sort by timestamp descending (most recent first)
            {'$sort': {'timestamp': -1}},
            
            # Join with surveys collection to get provider info
            {'$lookup': {
                'from': 'surveys',
                'localField': 'survey_id',
                'foreignField': '_id',
                'as': 'survey'
            }},
            
            # Unwind the survey array
            {'$unwind': {'path': '$survey', 'preserveNullAndEmptyArrays': False}},
            
            # Filter for CINT provider only
            {'$match': {'survey.provider': 'CINT'}},
            
            # Limit to requested number of results
            {'$limit': limit}
        ]
        
        cint_logs = list(allocation_log.aggregate(pipeline))
        
        # Build detailed payload information for each log
        for log in cint_logs:
            survey_id = log.get('survey_id')
            respondent_id = log.get('rid') or log.get('respondent_id')
            
            # Get survey details from the joined data
            survey = log.get('survey')
            external_survey_id = survey.get('external_id') if survey else None
            survey_name = survey.get('name') if survey else None
            
            # Get respondent details
            respondent = respondents_collection.find_one({'rid': respondent_id})
            
            # Try to get Cint survey details
            cint_survey = None
            if external_survey_id:
                cint_survey = cint_surveys.find_one({'survey_id': int(external_survey_id)})
            
            # Reconstruct the payload that was sent to Cint
            # Based on backend/routers/traffic.py lines 821-827
            payload = {
                'survey_id': str(external_survey_id) if external_survey_id else str(survey_id),
                'supplier_code': os.getenv('CINT_SUPPLIER_CODE', '6777'),
                'respondent_id': respondent_id,
                'secure_hash': '[HMAC-SHA256 hash - not stored]',
                'return_url': os.getenv('CINT_CALLBACK_URL', 'https://torpedo.cogentixresearch.com/api/cint/status'),
            }
            
            # Additional metadata
            metadata = {
                'timestamp': format_timestamp(log.get('timestamp')),
                'allocation_id': str(log.get('allocation_id', '')),
                'vid': log.get('vid'),
                'cc': log.get('cc'),
                'ip_address': log.get('ip_address'),
                'user_agent': log.get('user_agent'),
                'survey_name': survey_name,
                'survey_status': survey.get('status') if survey else None,
            }
            
            # Combine payload and metadata
            full_record = {
                'payload_sent_to_cint': payload,
                'metadata': metadata,
            }
            
            # Add Cint survey details if available
            if cint_survey:
                full_record['cint_survey_details'] = {
                    'survey_name': cint_survey.get('survey_name'),
                    'country_language': cint_survey.get('country_language'),
                    'loi': cint_survey.get('bid_length_of_interview'),
                    'cpi': cint_survey.get('revenue_per_interview', {}).get('value') if cint_survey.get('revenue_per_interview') else None,
                    'conversion': cint_survey.get('conversion'),
                }
            
            payloads.append(full_record)
        
        return payloads
        
    except Exception as e:
        print(f"Error retrieving payloads: {e}")
        import traceback
        traceback.print_exc()
        return []
    finally:
        client.close()

def print_payloads(payloads: List[Dict[str, Any]]):
    """Pretty print the payloads"""
    if not payloads:
        print("No Cint payloads found.")
        return
    
    print("=" * 80)
    print(f"LAST {len(payloads)} RESPONDENT PAYLOADS SENT TO CINT")
    print("=" * 80)
    print()
    
    for i, record in enumerate(payloads, 1):
        print(f"[{i}] Timestamp: {record['metadata']['timestamp']}")
        print("-" * 80)
        
        print("\nPayload sent to Cint API:")
        print("  POST https://api.samplicio.us/supply/v1/entrylinks")
        payload = record['payload_sent_to_cint']
        for key, value in payload.items():
            print(f"    {key}: {value}")
        
        print("\nMetadata:")
        metadata = record['metadata']
        for key, value in metadata.items():
            if key != 'timestamp':  # Already shown above
                print(f"    {key}: {value}")
        
        if 'cint_survey_details' in record:
            print("\nCint Survey Details:")
            details = record['cint_survey_details']
            for key, value in details.items():
                print(f"    {key}: {value}")
        
        print("\n" + "=" * 80)
        print()

def main():
    """Main execution"""
    import argparse
    
    parser = argparse.ArgumentParser(
        description='Retrieve last N respondent payloads sent to Cint',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  # Get last 10 payloads (default)
  python scripts/get_last_cint_payloads.py
  
  # Get last 20 payloads
  python scripts/get_last_cint_payloads.py --limit 20
  
  # Get payloads and save to JSON file
  python scripts/get_last_cint_payloads.py --json output.json
  
  # Run on VM via SSH
  ssh user@vm "cd /path/to/campaign_platform && python scripts/get_last_cint_payloads.py"
        """
    )
    
    parser.add_argument(
        '--limit',
        type=int,
        default=10,
        help='Number of payloads to retrieve (default: 10)'
    )
    
    parser.add_argument(
        '--json',
        type=str,
        help='Save output to JSON file instead of printing'
    )
    
    args = parser.parse_args()
    
    print(f"Connecting to MongoDB...")
    print(f"Retrieving last {args.limit} Cint respondent payloads...\n")
    
    payloads = get_last_cint_payloads(limit=args.limit)
    
    if args.json:
        # Save to JSON file
        with open(args.json, 'w') as f:
            json.dump(payloads, f, indent=2, default=str)
        print(f"Payloads saved to {args.json}")
    else:
        # Print to console
        print_payloads(payloads)
    
    print(f"\nTotal payloads retrieved: {len(payloads)}")

if __name__ == '__main__':
    main()
